zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/AnonymousAuthorizationTest.java

```diff
@@ -0,0 +1,184 @@
+          .endEvent()
+          .done();
+
+  private static final BpmnModelInstance PROCESS_WITH_NOT_EXISTING_FORM =
+      Bpmn.createExecutableProcess(PROCESS_ID_WITH_NOT_EXISTING_FORM)
+          .startEvent()
+          .userTask("task")
+          .zeebeFormId(FORM_ID_1)
+          .endEvent()
+          .done();
+
+  @Rule
+  public final RecordingExporterTestWatcher recordingExporterTestWatcher =
+      new RecordingExporterTestWatcher();
+
+  @BeforeClass
+  public static void setUp() {
+    final var username = UUID.randomUUID().toString();
+    userKey = ENGINE.user().newUser(username).create().getValue().getUserKey();
+    final var tenantKey = ENGINE.tenant().newTenant().withTenantId(TENANT).create().getKey();
+    ENGINE
+        .tenant()
+        .addEntity(tenantKey)
+        .withEntityKey(userKey)
+        .withEntityType(EntityType.USER)
+        .add();
+
+    ENGINE
+        .authorization()
+        .permission()
+        .withPermission(PermissionType.CREATE, "*")
+        .withResourceType(AuthorizationResourceType.DEPLOYMENT)
+        .withOwnerKey(userKey)
+        .withOwnerType(AuthorizationOwnerType.USER)
+        .add();
+
+    ENGINE.deployment().withXmlResource(PROCESS).deploy(userKey);
+  }
+
+  @Test
+  public void shouldCreateProcessInstanceWithAnonymousClaim() {
+    // given
+    final var anonymous =
+        AuthorizationUtil.getAuthInfoWithClaim(Authorization.AUTHORIZED_ANONYMOUS_USER, true);
+
+    // when
+    final var processInstanceKey =
+        ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create(anonymous);
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .exists())
+        .isTrue();
+  }
+
+  @Test
+  public void shouldNotCreateProcessInstanceWhenDisablingAnonymousClaim() {
+    // given
+    final var anonymousWithExplicitlyDisabledAnonymous =
+        AuthorizationUtil.getAuthInfoWithClaim(Authorization.AUTHORIZED_ANONYMOUS_USER, false);
+
+    // when
+    ENGINE
+        .processInstance()
+        .ofBpmnProcessId(PROCESS_ID)
+        .expectRejection()
+        .create(anonymousWithExplicitlyDisabledAnonymous);
+
+    // then
+    assertThat(RecordingExporter.records().withRejectionType(RejectionType.UNAUTHORIZED).exists())
+        .isTrue();
+  }
+
+  @Test
+  public void shouldNotCreateProcessInstanceWithoutAnonymousClaim() {
+    // given
+    final var anonymousWithExplicitlyDisabledAnonymous =
+        AuthorizationUtil.getAuthInfoWithClaim("foo", "bar");
+
+    // when
+    ENGINE
+        .processInstance()
+        .ofBpmnProcessId(PROCESS_ID)
+        .expectRejection()
+        .create(anonymousWithExplicitlyDisabledAnonymous);
+
+    // then
+    assertThat(RecordingExporter.records().withRejectionType(RejectionType.UNAUTHORIZED).exists())
+        .isTrue();
+  }
+
+  @Test
+  public void shouldResolveWithAnonymousClaim() {
```

Comment:
❓ What's the scenario here? An authorized user started an instance with a form start event referencing a non-existing form, so it ran into an incident. Now an anonymous user must be able to resolve this incident? Why should an anonymous user be able to do this? It doesn't seem safe because if there's an external endpoint to resolve an incident there's no mechanism stopping users with malicious intent to try and resolve incidents not related to the form start event. It's not limited to this sub-set of incidents.

Reply:
The scenario is that I — in the gateway (including the Tasklist v1 layer) — would want to run any command anonymously. However, this does not imply that the gateway will do this in all cases. In the context of keeping support for the Tasklist v1 API, this feature is needed for

* [Creating an instance via the public endpoint/form](https://github.com/camunda/camunda/pull/25799/files#diff-03e110ae401f7669422a8f9e1f45e4e1f86f651eae65790eca4ff9e23277540aR70-R78)
* [Completing jobs whereby the authorization check happened already in the Gateway/Tasklist](https://github.com/camunda/camunda/pull/25799/files#diff-03e110ae401f7669422a8f9e1f45e4e1f86f651eae65790eca4ff9e23277540aR132-R143)

I would not want to implement some special handling only for these two cases; instead, I would want to implement a simple general concept that would allow the running of any command anonymously. But again, by that, it does not mean that this will happen with all commands. Also, with that concept, the Gateway must explicitly "do something" to run a command anonymously; there is no implicit handling that would cause an anonymous call.

Additionally, the same adjustments as done with Tasklist should happen with Operate as well (if there is enough time left). There, the old Operate backend will do the checks (to verify the authenticated user has all the authorizations to perform that action) and then asynchronously, for example, resolve incidents (thereby, it would also make broker requests). In that case, I would also need this functionality instead of implementing special handling for the type of commands it will do.

In C7, there is a similar concept that allows me to run any type of command unauthenticated, hence without authorization and tenant check.

Just to repeat and to avoid confusion, it is not about having all endpoints open so that unauthenticated users can do malicious things. That's not the intention. It is more about having a simple concept in the engine, that allows the Gateway to execute commands anonymously. The contract is that the Gateway must ensure the authentication of the user and authorization, etc., is given in such cases, if possible (obviously, in the case of creating instances publicly, it is not possible, but in another case, it will do).

We have a similar concept in the C8 search stack, where the implementation must first fetch the entity before it can perform a proper authorization check based on it. There the search also happens explicitly without authentication.

That said, this test case does not imply that there will be an open endpoint that allows any unauthenticated user to resolve incidents. I wrote a test case to debug the code and ensure that it does not break when there is an anonymous call. So, I just submitted the test case instead of deleting it because it would not harm if it were there.

If you want me to delete that test case, I am happy to do so. Just let me know.

Reply:
Thanks for the additional explanation. It makes more sense to me now. So it's not that anyone can just resolve the incident, it's testing that the mechanism that commands can "skip" authorization checks if we tell it to. That makes sense to me 👍